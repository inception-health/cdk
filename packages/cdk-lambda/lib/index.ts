export { Lambda } from "./lambda";
export type { LambdaProps } from "./lambda";

export { DeadLetterQueue } from "./dlq/queue";
export type { DeadLetterQueueProps } from "./dlq/queue";

export { StandardNetwork } from "./network/network";
export type { StandardNetworkProps } from "./network/network";
