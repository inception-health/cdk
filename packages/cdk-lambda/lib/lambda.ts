import {
  aws_iam,
  aws_kms,
  aws_lambda,
  aws_lambda_nodejs,
  aws_secretsmanager,
  aws_logs,
  Duration,
  Size,
} from "aws-cdk-lib";
import { Construct } from "constructs";

import { DeadLetterQueue } from "./dlq/queue";

/**
 * Interface for LambdaProps
 */
export interface LambdaProps {
  /**
   * Activate a Dead Letter Queue for the lambda function
   *
   * @defaultValue false
   */
  deadLetterQueue?: boolean;

  /**
   * Description of the lambda function
   */
  description?: string;

  /**
   * Encryption key to be used
   */
  encryption?: aws_kms.IKey;

  /**
   * Environment variables for the lambda function
   */
  environment?: { [key: string]: string };

  /**
   * File containing the lambda function
   */
  file: string;

  /**
   * Handler for the lambda function
   *
   * @defaultValue "handler"
   */
  handler?: string;

  /**
   * Log retention period in days
   *
   * @defaultValue 2 months
   */
  logRetention?: aws_logs.RetentionDays;

  /**
   * Logging level
   *
   * @defaultValue INFO
   */
  loggingLevel?: "INFO" | "DEBUG" | "ERROR";

  /**
   * Memory size of the Lambda function in MB.
   *
   * @defaultValue `128` (MB)
   */
  memorySize?: number;

  /**
   * Name of the lambda function
   */
  name: string;

  /**
   * Ephemeral storage for the Lambda function.
   *
   * @defaultValue 512 MB
   */
  storage?: Size;

  /**
   * Timeout of the Lambda function.
   *
   * @defaultValue 30 seconds
   */
  timeout?: Duration;
}

const defaultProps: Partial<LambdaProps> = {
  handler: "handler",
  logRetention: aws_logs.RetentionDays.TWO_MONTHS,
  timeout: Duration.seconds(30),
  memorySize: 128,
  loggingLevel: "INFO",
  storage: Size.mebibytes(512),
  deadLetterQueue: false,
};

export interface LambdaCDK {
  encryption: aws_kms.IKey | undefined;
  function: aws_lambda.Function;
  logGroup: aws_logs.LogGroup;
  role: aws_iam.Role;
  secret: aws_secretsmanager.ISecret | undefined;
}

/**
 * `Lambda` is a class that represents an AWS Lambda function in a CDK
 * application.
 *
 * By default, the log group is encrypted with the key provided in the
 * `encryption` prop. If no key is provided, the log group is encrypted with the
 * default key provided by AWS. The log retention is set to 1 year by default.
 *
 * @example
 * ```typescript
 * const myLambda = new Lambda({
 *   name: 'myLambdaFunction',
 *   handler: 'index.handler',
 *   secretName: 'mySecret',
 * });
 * ```
 *
 * @remarks
 * - **Logging**: The lambda function logs to CloudWatch Logs. The Lambda
 *   captures logs in JSON structured format without having to use your own
 *   logging libraries. JSON structured logs make it easier to search, filter,
 *   and analyze large volumes of log entries. For more information, read about
 *   the new [AWS Lambda advanced
 *   logging](https://aws.amazon.com/blogs/compute/introducing-advanced-logging-controls-for-aws-lambda-functions/).
 *
 */
export class Lambda extends Construct implements aws_iam.IGrantable {
  public readonly cdk: LambdaCDK;

  public readonly dlq: {
    processor: Lambda | undefined;
    queue: DeadLetterQueue | undefined;
  };

  private readonly props: LambdaProps;

  constructor(scope: Construct, id: string, props: LambdaProps) {
    super(scope, id);

    // Initialize the properties
    this.props = Object.assign({}, defaultProps, props);
    this.cdk = {} as LambdaCDK;
    this.dlq = {} as Lambda["dlq"];

    // Setup
    this.setupLogGroup();
    this.setupRole();
    this.setupDlq();
    this.setupFunction();
    this.setupEncryption();
    this.setupDlqProcessor();
  }

  // !Public
  public get grantPrincipal(): aws_iam.IPrincipal {
    return this.cdk.role;
  }

  public addEnvironment(key: string, value: string) {
    this.cdk.function.addEnvironment(key, value);
  }

  public grantInvoke(identity: aws_iam.IGrantable): aws_iam.Grant {
    return this.cdk.function.grantInvoke(identity);
  }

  // !Setup
  private setupLogGroup() {
    this.cdk.logGroup = new aws_logs.LogGroup(this, "log-group", {
      logGroupName: `/aws/lambda/${this.props.name}`,
      encryptionKey: this.props.encryption,
      retention: this.props.logRetention,
    });
  }

  private setupRole() {
    this.cdk.role = new aws_iam.Role(this, "role", {
      assumedBy: new aws_iam.ServicePrincipal("lambda.amazonaws.com"),
    });
    this.cdk.role.addManagedPolicy(
      aws_iam.ManagedPolicy.fromAwsManagedPolicyName(
        "service-role/AWSLambdaBasicExecutionRole",
      ),
    );
  }

  private setupDlq() {
    if (!this.props.deadLetterQueue) return;

    this.dlq.queue = new DeadLetterQueue(this, "dlq", {
      encryption: this.props.encryption,
      name: this.props.name,
    });
  }

  private setupFunction() {
    this.cdk.function = new aws_lambda_nodejs.NodejsFunction(this, "function", {
      // Default
      architecture: aws_lambda.Architecture.ARM_64,
      tracing: aws_lambda.Tracing.ACTIVE,
      runtime: aws_lambda.Runtime.NODEJS_20_X,
      // Role
      role: this.cdk.role,
      // Performance
      timeout: this.props.timeout,
      memorySize: this.props.memorySize,
      ephemeralStorageSize: this.props.storage,
      // CloudWatch Lambda Insights is a monitoring and troubleshooting
      // solution for serverless applications running on AWS Lambda. The
      // solution collects, aggregates, and summarizes system-level metrics
      // including CPU time, memory, disk, and network. It also collects,
      // aggregates, and summarizes diagnostic information such as cold starts
      // and Lambda worker shutdowns to help you isolate issues with your Lambda
      // functions and resolve them quickly.
      // Ref: https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/Lambda-Insights.html
      insightsVersion: aws_lambda.LambdaInsightsVersion.VERSION_1_0_229_0,
      // Logs
      logGroup: this.cdk.logGroup,
      logFormat: aws_lambda.LogFormat.JSON,
      applicationLogLevel: aws_lambda.ApplicationLogLevel.INFO,
      systemLogLevel: aws_lambda.SystemLogLevel.INFO,
      // Customization
      functionName: this.props.name,
      description: this.props.description,
      // Environment and Encryption
      environmentEncryption: this.props.encryption,
      environment: this.props.environment,
      // Function
      entry: this.props.file,
      handler: this.props.handler,
      // Bundling
      bundling: {
        minify: true,
        sourceMap: false,
        externalModules: ["@aws-sdk/*"],
      },
    });
  }

  private setupEncryption() {
    if (!this.props.encryption) return;

    // Setup encryption
    this.cdk.encryption = this.props.encryption;
    this.cdk.encryption.grantDecrypt(this.cdk.function);
  }

  private setupDlqProcessor() {
    if (!this.props.deadLetterQueue) return;

    if (!this.dlq.queue) throw new Error("DLQ is undefined");

    this.dlq.processor = this.dlq.queue.attachRedrive(this);
  }
}
