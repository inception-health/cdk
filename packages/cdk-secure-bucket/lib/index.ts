export * from "./s3.secure";
export * from "./lifecycle";
