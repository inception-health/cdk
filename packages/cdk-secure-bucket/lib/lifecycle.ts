import { aws_s3, Duration } from "aws-cdk-lib";

// The lifecycle policy for items should be managed cohesively
export class LifecycleMode {
  static readonly NONE = new LifecycleMode();

  static readonly BASIC = new LifecycleMode({
    enabled: true,
    transitions: [
      {
        storageClass: aws_s3.StorageClass.INFREQUENT_ACCESS,
        transitionAfter: Duration.days(360),
      },
      {
        storageClass: aws_s3.StorageClass.GLACIER,
        transitionAfter: Duration.days(360 * 3),
      },
    ],
  });

  static readonly INTELLIGENT = new LifecycleMode({
    enabled: true,
    transitions: [
      {
        storageClass: aws_s3.StorageClass.INTELLIGENT_TIERING,
        transitionAfter: Duration.days(360),
      },
      {
        storageClass: aws_s3.StorageClass.GLACIER,
        transitionAfter: Duration.days(360 * 3),
      },
    ],
  });

  private constructor(public readonly rule?: aws_s3.LifecycleRule) {}
}
