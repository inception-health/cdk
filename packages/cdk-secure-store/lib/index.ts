export type { SecureStoreProps } from "./store";
export { SecureStore } from "./store";
