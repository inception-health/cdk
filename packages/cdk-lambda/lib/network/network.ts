import * as cdk from "aws-cdk-lib";
import {
  Vpc,
  SubnetType,
  SecurityGroup,
  Peer,
  Port,
  IVpc,
  ISecurityGroup,
  FlowLog,
  FlowLogTrafficType,
  FlowLogDestination,
} from "aws-cdk-lib/aws-ec2";
import { ILogGroup } from "aws-cdk-lib/aws-logs";
import { Construct } from "constructs";

import { Lambda } from "../lambda";

export interface StandardNetworkProps {
  encryption: cdk.aws_kms.IKey;
  maxAzs?: number;
  name?: string;
}

export interface StandardNetworkCdk {
  flowLogGroup: ILogGroup;
  flowLogRole: cdk.aws_iam.IRole;
  flowLogs: FlowLog;
  lambdaSecurityGroup: ISecurityGroup;
  vpc: IVpc;
}

const defaultProps: Partial<StandardNetworkProps> = {
  // Ensure high availability by spreading across 3 availability zones
  maxAzs: 3,
};

export class StandardNetwork extends Construct {
  public readonly cdk: StandardNetworkCdk;

  private readonly props: StandardNetworkProps;

  constructor(scope: Construct, id: string, props?: StandardNetworkProps) {
    super(scope, id);

    // Initialize the properties
    this.cdk = {} as StandardNetworkCdk;
    this.props = Object.assign({}, defaultProps, props);

    // Create the VPC
    this.cdk.vpc = new Vpc(this, "vpc", {
      maxAzs: this.props.maxAzs,
      subnetConfiguration: [
        {
          cidrMask: 24,
          name: "PublicSubnet",
          subnetType: SubnetType.PUBLIC,
          // (Control IDs: 164.308(a)(3)(i), 164.308(a)(4)(ii)(A),
          // 164.308(a)(4)(ii)(C), 164.312(a)(1), 164.312(e)(1)). Manage access
          // to the AWS Cloud by ensuring Amazon Virtual Private Cloud (VPC)
          // subnets are not automatically assigned a public IP address.
          mapPublicIpOnLaunch: false,
        },
        {
          cidrMask: 24,
          name: "PrivateSubnet",
          subnetType: SubnetType.PRIVATE_WITH_EGRESS,
        },
        {
          cidrMask: 28,
          name: "IsolatedSubnet",
          subnetType: SubnetType.PRIVATE_ISOLATED,
        },
      ],
    });

    // Create a security group for the Lambda function
    this.cdk.lambdaSecurityGroup = new SecurityGroup(
      this,
      "LambdaSecurityGroup",
      {
        vpc: this.cdk.vpc,
        description: "Security group for Lambda function",
        allowAllOutbound: true,
      },
    );

    // Restrict inbound traffic to the security group
    this.cdk.lambdaSecurityGroup.addIngressRule(
      Peer.anyIpv4(),
      Port.tcp(443),
      "Allow HTTPS traffic",
    );
    this.cdk.lambdaSecurityGroup.addIngressRule(
      Peer.anyIpv4(),
      Port.tcp(80),
      "Allow HTTP traffic",
    );

    // Create a log group for the VPC flow logs
    this.cdk.flowLogGroup = new cdk.aws_logs.LogGroup(this, "FlowLogGroup", {
      logGroupName: `/aws/vpc/flow-logs/${this.cdk.vpc.vpcId}`,
      retention: cdk.aws_logs.RetentionDays.ONE_MONTH,
      encryptionKey: this.props.encryption,
    });

    // Create a flow log for the VPC
    this.cdk.flowLogs = this.cdk.vpc.addFlowLog("FlowLog", {
      trafficType: FlowLogTrafficType.ALL,
      destination: FlowLogDestination.toCloudWatchLogs(this.cdk.flowLogGroup),
    });

    // Output the VPC ID and security group ID
    new cdk.CfnOutput(this, "VpcId", {
      value: this.cdk.vpc.vpcId,
    });

    new cdk.CfnOutput(this, "LambdaSecurityGroupId", {
      value: this.cdk.lambdaSecurityGroup.securityGroupId,
    });
  }

  public get vpc(): IVpc {
    return this.cdk.vpc;
  }

  public get lambdaSecurityGroup(): ISecurityGroup {
    return this.cdk.lambdaSecurityGroup;
  }

  public attachLambda(lambda: Lambda) {
    // lambda.cdk.function.connections.allowTo(
    //   this.cdk.lambdaSecurityGroup,
    //   Port.allTraffic(),
    //   "Allow Lambda to access resources in the VPC",
    // );
  }
}
