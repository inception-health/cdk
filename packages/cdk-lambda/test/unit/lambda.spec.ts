import path from "path";

import {
  App,
  Aspects,
  aws_dynamodb,
  aws_kms,
  aws_logs,
  CfnElement,
  Stack,
  aws_ec2,
} from "aws-cdk-lib";
import { Match, Template, Annotations } from "aws-cdk-lib/assertions";
import { HIPAASecurityChecks } from "cdk-nag";

import { Lambda } from "../../lib/lambda";
import { StandardNetwork } from "../../lib/network/network";

describe("Lambda", () => {
  let oldEnv: any;

  beforeAll(() => {
    oldEnv = process.env;
    process.env = { ...oldEnv, CDK_DEFAULT_REGION: "us-east-2" };
  });

  afterAll(() => {
    process.env = oldEnv;
  });

  describe("General", () => {
    it("should match the snapshot", () => {
      // Given
      const stack = new Stack();

      // When
      new Lambda(stack, "lambda-demo", {
        name: "hello-world",
        file: path.join(__dirname, "./mocks/function.ts"),
      });

      // Then
      expect(Template.fromStack(stack)).toMatchSnapshot();
    });

    it.skip("should pass cdk-nag HIPAASecurityChecks configuration", async () => {
      // GIVEN
      const app = new App();
      const stack = new Stack(app);
      Aspects.of(stack).add(new HIPAASecurityChecks({ verbose: true }));

      // WHEN
      const network = new StandardNetwork(stack, "TestVpc", {
        name: "TestVpc",
        encryption: new aws_kms.Key(stack, "Key"),
      });
      new Lambda(stack, "lambda-demo", {
        name: "hello-world",
        file: path.join(__dirname, "./mocks/function.ts"),
        vpc: network.vpc,
        vpcSubnets: network.cdk.vpc.selectSubnets({
          subnetType: aws_ec2.SubnetType.PRIVATE_WITH_EGRESS,
        }),
      });
      app.synth();

      // THEN
      const errors = Annotations.fromStack(stack).findError(
        "*",
        Match.stringLikeRegexp("HIPAA.*"),
      );
      expect(errors).toHaveLength(0);

      const warnings = Annotations.fromStack(stack).findWarning(
        "*",
        Match.stringLikeRegexp("HIPAA.*"),
      );
      expect(warnings).toHaveLength(0);
    });

    it("should have a minimum build", () => {
      // Given
      const stack = new Stack();

      // When
      new Lambda(stack, "lambda-demo", {
        name: "hello-world",
        file: path.join(__dirname, "./mocks/function.ts"),
      });

      // Then
      const template = Template.fromStack(stack);
      template.resourceCountIs("AWS::Lambda::Function", 1);
    });

    it("should match a build with all configuration", () => {
      // Given
      const stack = new Stack();
      const key = new aws_kms.Key(stack, "key");

      // When
      new Lambda(stack, "lambda-demo", {
        name: "hello-world",
        description: "here is a demo description of the lambda",
        file: path.join(__dirname, "./mocks/function.ts"),
        handler: "handler2.ts",
        environment: {
          HELLO_WORLD_ENV: "test-demo-env",
        },
        encryption: key,
      });

      // Then
      const template = Template.fromStack(stack);
      template.resourceCountIs("AWS::Lambda::Function", 1);
      template.hasResourceProperties("AWS::Lambda::Function", {
        Description: "here is a demo description of the lambda",
        Handler: "handler2.ts",
        Environment: {
          Variables: {
            HELLO_WORLD_ENV: "test-demo-env",
          },
        },
        KmsKeyArn: Match.anyValue(),
      });
    });
  });

  describe("Lambda Concurrency", () => {
    it("should have a default reserved concurrency of 20", () => {
      // Given
      const stack = new Stack();

      // When
      new Lambda(stack, "lambda-demo", {
        name: "hello-world",
        file: path.join(__dirname, "./mocks/function.ts"),
      });

      // Then
      const template = Template.fromStack(stack);
      template.hasResourceProperties("AWS::Lambda::Function", {
        ReservedConcurrentExecutions: 20,
      });
    });

    it("should have a custom reserved concurrency", () => {
      // Given
      const stack = new Stack();

      // When
      new Lambda(stack, "lambda-demo", {
        name: "hello-world",
        file: path.join(__dirname, "./mocks/function.ts"),
        reservedConcurrency: 10,
      });

      // Then
      const template = Template.fromStack(stack);
      template.hasResourceProperties("AWS::Lambda::Function", {
        ReservedConcurrentExecutions: 10,
      });
    });
  });

  describe("Lambda Memory", () => {
    it("should have a default memory size of 128", () => {
      // Given
      const stack = new Stack();

      // When
      new Lambda(stack, "lambda-demo", {
        name: "hello-world",
        file: path.join(__dirname, "./mocks/function.ts"),
      });

      // Then
      const template = Template.fromStack(stack);
      template.hasResourceProperties("AWS::Lambda::Function", {
        MemorySize: 128,
      });
    });

    it("should have a custom memory size", () => {
      // Given
      const stack = new Stack();

      // When
      new Lambda(stack, "lambda-demo", {
        name: "hello-world",
        file: path.join(__dirname, "./mocks/function.ts"),
        memorySize: 256,
      });

      // Then
      const template = Template.fromStack(stack);
      template.hasResourceProperties("AWS::Lambda::Function", {
        MemorySize: 256,
      });
    });
  });

  describe("Lambda Logs", () => {
    it("should have a log group associated with the lambda", () => {
      // Given
      const stack = new Stack();

      // When
      const lambda = new Lambda(stack, "lambda-demo", {
        name: "hello-world",
        file: path.join(__dirname, "./mocks/function.ts"),
      });

      // Then
      const template = Template.fromStack(stack);
      const expectedRefLog = stack.getLogicalId(
        lambda.cdk.logGroup.node.defaultChild as CfnElement,
      );

      template.hasResourceProperties("AWS::Lambda::Function", {
        LoggingConfig: {
          LogGroup: {
            Ref: expectedRefLog,
          },
        },
      });
    });

    it("should have logs with a default retention period", () => {
      // Given
      const stack = new Stack();

      // When
      new Lambda(stack, "lambda-demo", {
        name: "hello-world",
        file: path.join(__dirname, "./mocks/function.ts"),
      });

      // Then
      const template = Template.fromStack(stack);
      template.hasResourceProperties("AWS::Logs::LogGroup", {
        RetentionInDays: 60,
      });
    });

    it("should have logs with a custom retention period", () => {
      // Given
      const stack = new Stack();

      // When
      new Lambda(stack, "lambda-demo", {
        name: "hello-world",
        file: path.join(__dirname, "./mocks/function.ts"),
        logRetention: aws_logs.RetentionDays.ONE_WEEK,
      });

      // Then
      const template = Template.fromStack(stack);
      template.hasResourceProperties("AWS::Logs::LogGroup", {
        RetentionInDays: 7,
      });
    });

    it("should have logs formatted as JSON", () => {
      // Given
      const stack = new Stack();

      // When
      new Lambda(stack, "lambda-demo", {
        name: "hello-world",
        file: path.join(__dirname, "./mocks/function.ts"),
      });

      // Then
      const template = Template.fromStack(stack);
      template.hasResourceProperties("AWS::Lambda::Function", {
        LoggingConfig: {
          LogFormat: "JSON",
        },
      });
    });

    it("should have INFO logging level by default", () => {
      // Given
      const stack = new Stack();

      // When
      new Lambda(stack, "lambda-demo", {
        name: "hello-world",
        file: path.join(__dirname, "./mocks/function.ts"),
      });

      // Then
      const template = Template.fromStack(stack);
      template.hasResourceProperties("AWS::Lambda::Function", {
        LoggingConfig: {
          ApplicationLogLevel: "INFO",
          SystemLogLevel: "INFO",
        },
      });
    });
  });

  describe("Lambda Logs", () => {
    it("should have a role associated with the lambda", () => {
      // Given
      const stack = new Stack();

      // When
      const lambda = new Lambda(stack, "lambda-demo", {
        name: "hello-world",
        file: path.join(__dirname, "./mocks/function.ts"),
      });

      // Then
      const template = Template.fromStack(stack);
      const expectedRefRole = stack.getLogicalId(
        lambda.cdk.role.node.defaultChild as CfnElement,
      );

      template.hasResourceProperties("AWS::Lambda::Function", {
        Role: {
          "Fn::GetAtt": [expectedRefRole, "Arn"],
        },
      });

      template.hasResourceProperties("AWS::IAM::Role", {
        AssumeRolePolicyDocument: {
          Statement: [
            {
              Action: "sts:AssumeRole",
              Effect: "Allow",
              Principal: {
                Service: "lambda.amazonaws.com",
              },
            },
          ],
          Version: "2012-10-17",
        },
        ManagedPolicyArns: [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  Ref: "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  Ref: "AWS::Partition",
                },
                ":iam::aws:policy/CloudWatchLambdaInsightsExecutionRolePolicy",
              ],
            ],
          },
        ],
      });
    });

    it("should be able to grant permissions to a role", () => {
      // Given
      const stack = new Stack();
      const lambda = new Lambda(stack, "lambda-demo", {
        name: "hello-world",
        file: path.join(__dirname, "./mocks/function.ts"),
      });
      const table = new aws_dynamodb.TableV2(stack, "store-demo", {
        partitionKey: { name: "id", type: aws_dynamodb.AttributeType.STRING },
      });

      // When
      table.grantReadData(lambda);

      // Then
      const template = Template.fromStack(stack);
      const expectedRefTable = stack.getLogicalId(
        table.node.defaultChild as CfnElement,
      );

      template.hasResourceProperties("AWS::IAM::Policy", {
        PolicyDocument: {
          Statement: Match.arrayWith([
            {
              Action: [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:DescribeTable",
              ],
              Effect: "Allow",
              Resource: {
                "Fn::GetAtt": [expectedRefTable, "Arn"],
              },
            },
          ]),
          Version: "2012-10-17",
        },
      });
    });
  });
});
