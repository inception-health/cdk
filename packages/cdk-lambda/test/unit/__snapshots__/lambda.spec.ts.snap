// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lambda General should match the snapshot 1`] = `
{
  "Mappings": {
    "CloudwatchlambdainsightsversionMap": {
      "af-south-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:af-south-1:012438385374:layer:LambdaInsightsExtension-Arm64:18",
      },
      "ap-east-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:ap-east-1:519774774795:layer:LambdaInsightsExtension-Arm64:18",
      },
      "ap-northeast-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:ap-northeast-1:580247275435:layer:LambdaInsightsExtension-Arm64:31",
      },
      "ap-northeast-2": {
        "1x0x333x0xarm64": "arn:aws:lambda:ap-northeast-2:580247275435:layer:LambdaInsightsExtension-Arm64:19",
      },
      "ap-northeast-3": {
        "1x0x333x0xarm64": "arn:aws:lambda:ap-northeast-3:194566237122:layer:LambdaInsightsExtension-Arm64:17",
      },
      "ap-south-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:ap-south-1:580247275435:layer:LambdaInsightsExtension-Arm64:22",
      },
      "ap-south-2": {
        "1x0x333x0xarm64": "arn:aws:lambda:ap-south-2:891564319516:layer:LambdaInsightsExtension-Arm64:6",
      },
      "ap-southeast-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:ap-southeast-1:580247275435:layer:LambdaInsightsExtension-Arm64:20",
      },
      "ap-southeast-2": {
        "1x0x333x0xarm64": "arn:aws:lambda:ap-southeast-2:580247275435:layer:LambdaInsightsExtension-Arm64:20",
      },
      "ap-southeast-3": {
        "1x0x333x0xarm64": "arn:aws:lambda:ap-southeast-3:439286490199:layer:LambdaInsightsExtension-Arm64:18",
      },
      "ca-central-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:ca-central-1:580247275435:layer:LambdaInsightsExtension-Arm64:18",
      },
      "eu-central-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:eu-central-1:580247275435:layer:LambdaInsightsExtension-Arm64:20",
      },
      "eu-north-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:eu-north-1:580247275435:layer:LambdaInsightsExtension-Arm64:18",
      },
      "eu-south-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:eu-south-1:339249233099:layer:LambdaInsightsExtension-Arm64:18",
      },
      "eu-south-2": {
        "1x0x333x0xarm64": "arn:aws:lambda:eu-south-2:352183217350:layer:LambdaInsightsExtension-Arm64:6",
      },
      "eu-west-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:eu-west-1:580247275435:layer:LambdaInsightsExtension-Arm64:20",
      },
      "eu-west-2": {
        "1x0x333x0xarm64": "arn:aws:lambda:eu-west-2:580247275435:layer:LambdaInsightsExtension-Arm64:20",
      },
      "eu-west-3": {
        "1x0x333x0xarm64": "arn:aws:lambda:eu-west-3:580247275435:layer:LambdaInsightsExtension-Arm64:18",
      },
      "me-south-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:me-south-1:285320876703:layer:LambdaInsightsExtension-Arm64:18",
      },
      "sa-east-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:sa-east-1:580247275435:layer:LambdaInsightsExtension-Arm64:18",
      },
      "us-east-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:us-east-1:580247275435:layer:LambdaInsightsExtension-Arm64:20",
      },
      "us-east-2": {
        "1x0x333x0xarm64": "arn:aws:lambda:us-east-2:580247275435:layer:LambdaInsightsExtension-Arm64:22",
      },
      "us-gov-east-1": {
        "1x0x333x0xarm64": "arn:aws-us-gov:lambda:us-gov-east-1:122132214140:layer:LambdaInsightsExtension-Arm64:3",
      },
      "us-gov-west-1": {
        "1x0x333x0xarm64": "arn:aws-us-gov:lambda:us-gov-west-1:751350123760:layer:LambdaInsightsExtension-Arm64:3",
      },
      "us-west-1": {
        "1x0x333x0xarm64": "arn:aws:lambda:us-west-1:580247275435:layer:LambdaInsightsExtension-Arm64:18",
      },
      "us-west-2": {
        "1x0x333x0xarm64": "arn:aws:lambda:us-west-2:580247275435:layer:LambdaInsightsExtension-Arm64:20",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "lambdademofunctionC779DFD0": {
      "DependsOn": [
        "lambdademoroleDefaultPolicy921C4C2D",
        "lambdademorole95F35F6F",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "7671fef3231da5b8460e43bcffeb2c9d9d7763730b7734f98ef6960dfeda58d3.zip",
        },
        "EphemeralStorage": {
          "Size": 512,
        },
        "FunctionName": "hello-world",
        "Handler": "index.handler",
        "Layers": [
          {
            "Fn::FindInMap": [
              "CloudwatchlambdainsightsversionMap",
              {
                "Ref": "AWS::Region",
              },
              "1x0x333x0xarm64",
            ],
          },
        ],
        "LoggingConfig": {
          "ApplicationLogLevel": "INFO",
          "LogFormat": "JSON",
          "LogGroup": {
            "Ref": "lambdademologgroup4B86AFE9",
          },
          "SystemLogLevel": "INFO",
        },
        "MemorySize": 128,
        "ReservedConcurrentExecutions": 20,
        "Role": {
          "Fn::GetAtt": [
            "lambdademorole95F35F6F",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 30,
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "lambdademologgroup4B86AFE9": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "LogGroupName": "/aws/lambda/hello-world",
        "RetentionInDays": 60,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "lambdademorole95F35F6F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/CloudWatchLambdaInsightsExecutionRolePolicy",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "lambdademoroleDefaultPolicy921C4C2D": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "lambdademoroleDefaultPolicy921C4C2D",
        "Roles": [
          {
            "Ref": "lambdademorole95F35F6F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
